tools/lxdclient/client_image.go

```diff
@@ -92,56 +97,107 @@ func (p *progressContext) copyProgress(progress string) {
 // @param copyProgressHandler: a callback function. If we have to download an
 // image, we will call this with messages indicating how much of the download
 // we have completed (and where we are downloading it from).
-func (i *imageClient) EnsureImageExists(series string, sources []Remote, copyProgressHandler func(string)) error {
+func (i *imageClient) EnsureImageExists(series string, sources []Remote, copyProgressHandler func(string)) (string, error) {
 	// TODO(jam) Find a way to test this, even though lxd.Client can't
 	// really be stubbed out because CopyImage takes one directly and pokes
 	// at private methods so we can't easily tweak it.
-	name := i.ImageNameForSeries(series)
 
-	lastErr := errors.New("image not imported!")
+	// NOTE(axw) architecture should be passed into this function, as
+	// indicated by the TODO in the function's doc comment. When it is,
+	// use get rid of this.
+	const arch = "amd64"
+
+	// First check if the image exists locally.
+	var lastErr error
+	imageName := seriesLocalAlias(series, arch)
+	target := i.raw.GetAlias(imageName)
+	if target != "" {
```

Comment:
> I go back and forth on this one. If we had implemented caching of images on the controller, I heavily prefer using the upstream alias, because I don't really want one machine that deployed a container a long time ago to have a separate notion of what it is from all the other machines.

I don't understand what the alias has to do with that. The alias used is orthogonal to image freshness. If you copy the upstream image into your LXD and give it the name "foo", you can still have it kept up-to-date by having the auto-update flag enabled.

> Aside from that, we have sort-of wanted the ability to sneak custom images into the workflow. So I'm reasonably happy to have some way for users to supply their own images, but doing it on each machine where you want to operate is a nightmare of distributed state, which is just ugly.

The only change in behaviour is that we will now use images that don't exist in a remote. We still set the auto-update flag, which means all the LXDs will still update the juju/foo/bar image as they did for the ubuntu-foo image.

> Is there a way to do this by 'provider', so that lxd provider prefers local aliases, but cloud images prefer official upstreams?

We could, but I can't see the value in the added complexity.